service/project.go

```diff
@@ -211,6 +219,147 @@ func (s *ProjectService) DeleteEnvironment(ctx context.Context, projectID, envID
 	return s.envRepo.Delete(ctx, envID)
 }
 
+func (s *ProjectService) CreateInvitation(ctx context.Context, c model.ProjectInvitationCreation, authUserID uuid.UUID) (model.ProjectInvitation, error) {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	if _, err := s.Get(ctx, c.ProjectID, authUserID); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	tkn, err := cryptox.NewToken()
+	if err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	i, err := model.NewProjectInvitation(c, tkn)
+	if err != nil {
+		return model.ProjectInvitation{}, apierrors.NewProjectInvitationUnprocessableError().Wrap(err).WithMessage(err.Error())
+	}
+
+	// TODO check if the user is already a member of the project
+
+	if exists, err := s.invitationExists(ctx, i.Email, c.ProjectID); err != nil {
+		return model.ProjectInvitation{}, err
+	} else if exists {
+		return model.ProjectInvitation{}, apierrors.NewProjectInvitationAlreadyExistsError()
+	}
+
+	if err := s.invitationRepo.Create(ctx, i); err != nil {
+		return model.ProjectInvitation{}, err
+	}
+
+	// TODO send invitation email
+
+	return i, nil
+}
+
+func (s *ProjectService) GetInvitations(ctx context.Context, projectID, authUserID uuid.UUID) ([]model.ProjectInvitation, error) {
+	if err := s.authSvc.AuthorizeAdminRole(ctx, authUserID); err != nil {
+		return nil, err
+	}
+
+	_, err := s.Get(ctx, projectID, authUserID)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := s.invitationRepo.ReadAllForProject(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := model.ValidateProjectInvitations(i); err != nil {
```

Comment:
Checking if invitations are returned in a valid state from the repository.

Reply:
Yes, it's true. Even my need to explain why I do validation in some places and not in others is not a good sign.